ctk/common/src/main/kotlin/org/codice/compliance/verification/binding/RedirectBindingVerifier.kt

```diff
@@ -129,74 +135,78 @@ class RedirectBindingVerifier(private val response: IdpRedirectResponseDecorator
-            }
+    private fun verifyNoNullsAndParse(): Map<String, String> {
+        val url = response.header(LOCATION) ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Url not found.")
+
+        val splitUrl = Splitter.on("?").splitToList(url)
+
+        splitUrl[0] ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Path not found.")
+
+        val parameters = splitUrl[1] ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "Parameters not found.")
+
+        val paramMap = parameters.split("&")
+                .map { s -> s.split("=") }
+                .associate { s -> s[0] to s[1] }
+
+        paramMap[SAML_RESPONSE] ?: throw SAMLComplianceException.create(
+                SAMLBindings_3_4_4_b,
+                message = "SAMLResponse not found.")
+
+        if (isRelayStateGiven && paramMap[RELAY_STATE] == null) {
+            throw SAMLComplianceException.create(
+                    SAMLBindings_3_4_3_b,
+                    message = "RelayState not found.")
         }
+
+        return paramMap
     }
 
     /**
      * Verifies the presence of redirect parameters according to the redirect binding rules in the
      * binding spec (Negative path)
      * 3.4.4 Message Encoding
      */
-    private fun verifyNoNullsErrorResponse() {
-        with(response) {
-            if (isUrlNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Url not found." +
-                                "\n$IDP_ERROR_RESPONSE_REMINDER_MESSAGE")
-            }
-            if (isPathNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Path not found." +
-                                "\n$IDP_ERROR_RESPONSE_REMINDER_MESSAGE")
-            }
-            if (isParametersNull) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "Parameters not found." +
-                                "\n$IDP_ERROR_RESPONSE_REMINDER_MESSAGE")
-            }
-            if (samlResponse == null) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_4_b,
-                        message = "SAMLResponse not found." +
-                                "\n$IDP_ERROR_RESPONSE_REMINDER_MESSAGE")
-            }
-            if (isRelayStateGiven && relayState == null) {
-                throw SAMLComplianceException.create(
-                        SAMLBindings_3_4_3_b,
-                        message = "RelayState not found." +
-                                "\n$IDP_ERROR_RESPONSE_REMINDER_MESSAGE")
-            }
+    private fun verifyNoNullsErrorAndParse(): Map<String, String> {
```

Comment:
✏️ This method looks very similar to its sister function. Should be able to refactor.

Reply:
It's identical except for tacking on additional info in every exception message in here. That's true of almost all of the "error" sister functions. I was trying to avoid doing a big refactor on that in this PR. I also didn't like any alternative solutions I could think of.

Reply:
It should definitely happen after. Could maybe pass in a messageBuilder function that generates a string from an input string? We can review later.